### ISSUE TITLE
manual/scalaGuide/main/json/ScalaJson.md

### ISSUE THREAD

**Issue by garbagetown** [playframework-ja/translation-project#615]
- https://github.com/garbagetown/playdocja/blob/2.2.0/documentation/2.2.0/manual/scalaGuide/main/json/ScalaJson.md

```
--- /Users/garbagetown/Desktop/2.1.5/manual/scalaGuide/main/json/ScalaJson.md   2013-09-19 22:53:02.000000000 +0900
+++ /Users/garbagetown/Desktop/2.2.0/manual/scalaGuide/main/json/ScalaJson.md   2013-09-19 10:11:22.000000000 +0900
@@ -4,13 +4,13 @@

 The recommended way of dealing with JSON is using Play’s typeclass based JSON library, located at '''play.api.libs.json'''. 

-For parsing JSON strings, Play uses super-fast Java based JSON library, [Jackson](http://jackson.codehaus.org/).  
+For parsing JSON strings, Play uses the super-fast Java based JSON library, [Jackson](http://jackson.codehaus.org/).

 The benefit of this approach is that both the Java and the Scala side of Play can share the same underlying library (Jackson), while Scala users can enjoy the extra type safety and functional aspects that Play’s JSON support brings to the table.

 ## JSON is an AST (_Abstract Syntax Tree_)

-Take JSON example:
+Take this JSON for example:

 '''json
 { 
@@ -28,7 +28,7 @@
 }
 '''

-This can be seen as a tree structure using the 2 following structures:
+This can be seen as a tree structure using the two following structures:

 - **JSON object** contains a set of `name` / `value` pairs:
     - `name` is a String
@@ -46,50 +46,53 @@

 ## Json Data Types

-`play.api.libs.json` package contains 7 JSON data types reflecting exactly the previous structure.
+The [`play.api.libs.json`](api/scala/index.html#play.api.libs.json.package) package contains the seven JSON data types, reflecting this structure.

-### '''JsObject''' 
+### [`JsObject`](api/scala/index.html#play.api.libs.json.JsObject)

-- This is a set of name/value pairs as described in standard.
-- `{ "name" : "toto", "age" : 45 }` as a JSON example.
+This is a set of name/value pairs as described in the standard, for example:

-### '''JsNull'''
+'''json
+{ "name" : "toto", "age" : 45 }
+'''
+
+### [`JsNull`](api/scala/index.html#play.api.libs.json.JsNull)
+
+This represents a `null` value in JSON.

-This represents `null` value in JSON
+### [`JsBoolean`](api/scala/index.html#play.api.libs.json.JsBoolean)

-### '''JsBoolean'''
+This is a boolean with a value of `true` or `false`.

-This is a boolean with value `true` or `false`
+### [`JsNumber`](api/scala/index.html#play.api.libs.json.JsNumber)

-### '''JsNumber'''
+JSON does NOT discriminate `short`, `int`, `long`, `float`, `double` and `BigDecimal`, so it is represented by a `JsNumber` containing a `BigDecimal`.  The Play JSON API brings more type precision when converting to Scala structures.

-- JSON does NOT discriminate `short`, `int`, `long`, `float`, `double`, `bigdecimal` so it is represented by `JsNumber` containing a `bigdecimal`. 
-- Play JSON API brings more type precision when converting to Scala structures.
+### [`JsArray`](api/scala/index.html#play.api.libs.json.JsArray)

-### '''JsArray''' 
+An array is a sequence of any Json value types (not necessarily the same type), for example:

-- An array is a sequence of any Json value types (not necessarily the same type).
-- `[ "alpha", "beta", true, 123.44, 334]` as a JSON example.
+'''
+[ "alpha", "beta", true, 123.44, 334]
+'''

-### '''JsString'''
+### [`JsString`](api/scala/index.html#play.api.libs.json.JsString)

 A classic String.

-### '''JsUndefined'''
+## Other data types
+
+### [`JsUndefined`](api/scala/index.html#play.api.libs.json.JsUndefined)

 This is not part of the JSON standard and is only used internally by the API to represent some error nodes in the AST.

-### '''JsValue'''
+### [`JsValue`](api/scala/index.html#play.api.libs.json.JsValue)

-All previous types inherit from the generic JSON trait, '''JsValue'''.
+This is the super type of all the other types.

-## Minimal Import to work with basic JSON API
+## Working with JSON

-'''scala
-import play.api.libs.json.Json
-'''
-
-This import give access to the most basic JSON features :
+The entry point into Play's JSON API is [`play.api.libs.json.Json`](api/scala/index.html#play.api.libs.json.Json$).  It provides the following methods:

 - `Json.parse` : parses a string to JsValue
 - `Json.stringify` : stringifies a JsValue using compact printer (NO line feed/indentation)
@@ -104,192 +107,65 @@

 You can easily parse any JSON string as a `JsValue`:

-'''
-import play.api.libs.json.Json
+@[parse-json](code/ScalaJson.scala)

-val json: JsValue = Json.parse("""
-{ 
-  "user": {
-    "name" : "toto",
-    "age" : 25,
-    "email" : "toto@jmail.com",
-    "isAlive" : true,
-    "friend" : {
-     "name" : "tata",
-     "age" : 20,
-     "email" : "tata@coldmail.com"
-    }
-  } 
-}
-""")
-'''
-
-This sample is used in all next samples.
-
-As explained previously, the parsing is performed by [Jackson](http://jackson.codehaus.org/).
+The `json` value that the result was assigned to is used in subsequent code samples below.

 ## Constructing JSON directly

 ### Raw way

-The previous sample Json object can be created in other ways too. 
-Here is the raw approach.
+The previous sample Json object can be created in other ways too.  Here is the raw approach:

-'''
-import play.api.libs.json._
-
-JsObject(
-  "users" -> JsArray(
-    JsObject(
-      "name" -> JsString("Bob") ::
-      "age" -> JsNumber(31) ::
-      "email" -> JsString("bob@gmail.com") ::
-      Nil) ::
-    JsObject(
-      "name" -> JsString("Kiki") ::
-      "age" -> JsNumber(25) ::
-      "email" -> JsNull ::
-      Nil
-    ) :: Nil
-  ) :: Nil
-)
-'''
+@[construct-json-case-class](code/ScalaJson.scala)

 ### Preferred way

-Play now provides a simplified syntax to build your JSON.
-The previous JsObject can be constructed as following:
-
-'''
-import play.api.libs.json.Json
-
-Json.obj(
-  "users" -> Json.arr(
-    Json.obj(
-      "name" -> "bob",
-      "age" -> 31,
-      "email" -> "bob@gmail.com"     
-    ),
-    Json.obj(
-      "name" -> "kiki",
-      "age" -> 25,
-      "email" -> JsNull      
-    )
-  )
-)
-'''
+Play also provides a simplified syntax to build your JSON.  The previous JsObject can be constructed using the following:

+@[construct-json-dsl](code/ScalaJson.scala)

 ## Serializing JSON

 Serializing a `JsValue` to its JSON String representation is easy:

-'''
-import play.api.libs.json.Json
-
-val jsonString: String = Json.stringify(jsValue)
-'''
-
+@[serialise-json](code/ScalaJson.scala)

 ## Accessing Path in a JSON tree 

-As soon as you have a `JsValue` you can navigate into the JSON tree.  
-The API looks like the one provided to navigate into XML document by Scala using `NodeSeq` except you retrieve `JsValue`.
+As soon as you have a `JsValue` you can navigate into the JSON tree.  The API looks like the one provided to navigate into XML document by Scala using `NodeSeq` except you retrieve `JsValue`.

 ### Simple path `\`

-'''scala
-// Here we import everything under json in case we need to manipulate different Json types
-scala> import play.api.libs.json._
+You can navigate through properties in an object using the `\` method:

-scala> val name: JsValue = json \ "user" \ "name"
-name: play.api.libs.json.JsValue = "toto"
-'''
+@[traverse-path](code/ScalaJson.scala)

 ### Recursive path `\\`
- 
-'''scala
-// recursively searches in the sub-tree and returns a Seq[JsValue]
-// of all found JsValue
-scala> val emails: Seq[String] = json \ "user" \\ "email"
-emails: Seq[play.api.libs.json.JsValue] = List("toto@jmail.com", "tata@coldmail.com")
-'''
- 
+
+@[recursive-traverse-path](code/ScalaJson.scala)

 ## Converting JsValue to Scala Value

-While navigating JSON tree, you retrieve `JsValue` but you may want to convert the JsValue to a Scala type. 
-For ex, a `JsString` to a `String` or a `JsNumber` to a `Long` (if it can be converted).
+While navigating JSON tree, you retrieve `JsValue`, however you may want to convert the JsValue to a Scala type.  For example, you may want a `JsString` to be converted to a `String`, or a `JsNumber` to be converted to a `Long`.

 ### Unsafe conversion with `json.as[T]`

-`as[T]` is unsafe because it tries to access the path and to convert to the required type. But if the path is not found or the conversion not possible, it generates a `JsResultException` RuntimeException containing detected errors.
-
-#### case OK: path found & conversion possible
-
-'''scala
-// returns the value converted to provided type (if possible and if found)
-scala> val name: String = (json \ "user" \ "name").as[String]
-name: String = toto
-'''
+The simplest way to convert it to a value is to use the `as[T]` method, like so:

-#### case KO: Path not found
+@[as-method](code/ScalaJson.scala)

-'''scala
-scala> val nameXXX: String = (json \ "user" \ "nameXXX").as[String]
-play.api.libs.json.JsResultException: JsResultException(errors:List((,List(ValidationError(validate.error.expected.jsstring,WrappedArray())))))
-   at play.api.libs.json.JsValue$$anonfun$4.apply(JsValue.scala:65)
-   at play.api.libs.json.JsValue$$anonfun$4.apply(JsValue.scala:65)
-   at play.api.libs.json.JsResult$class.fold(JsResult.scala:69)
-   at play.api.libs.json.JsError.fold(JsResult.scala:10)
-   at play.api.libs.json.JsValue$class.as(JsValue.scala:63)
-   at play.api.libs.json.JsUndefined.as(JsValue.scala:96)
-'''
+This method however is unsafe, if the path is not found, or the conversion is not possible, a `JsResultException` is thrown, containing the error.

 > Please note the error that doesn't return `path.not.found` as you may expect. This is a difference from JSON combinators presented later in the doc. 
 > This is due to the fact that `(json \ "user" \ "nameXXX")` returns `JsNull` and the implicit `Reads[String]` here awaits a `JsString` which explains the detected error.

-
-#### case KO: Conversion not possible
-
-'''scala
-scala> val name: Long = (json \ "user" \ "name").as[Long]
-play.api.libs.json.JsResultException: JsResultException(errors:List((,List(ValidationError(validate.error.expected.jsnumber,WrappedArray())))))
-   at play.api.libs.json.JsValue$$anonfun$4.apply(JsValue.scala:65)
-   at play.api.libs.json.JsValue$$anonfun$4.apply(JsValue.scala:65)
-   at play.api.libs.json.JsResult$class.fold(JsResult.scala:69)
-   at play.api.libs.json.JsError.fold(JsResult.scala:10)
-   at play.api.libs.json.JsValue$class.as(JsValue.scala:63)
-   at play.api.libs.json.JsString.as(JsValue.scala:111)
-'''
-
-<br/>
 ### Safer conversion with `Option[T]`

-`as[T]` is immediate but not robust so there is `asOpt[T]` which returns None in case of error of any type.
+The `asOpt[T]` method is like `as[T]`, however it will return `None` instead of throwing an exception if the path isn't found, or the conversion isn't possible:

-#### case OK: path found & conversion possible
+@[as-opt](code/ScalaJson.scala)

-'''scala
-scala> val maybeName: Option[String] = (json \ "user" \ "name").asOpt[String]
-maybeName: Option[String] = Some(toto)
-'''
-
-#### case KO: Path not found
-
-'''scala
-scala> val maybeNameXXX: Option[String] = (json \ "user" \ "nameXXX").asOpt[String]
-maybeNameXXX: Option[String] = None
-'''
-
-#### case KO: Conversion not possible
-
-'''scala
-scala> val maybeNameLong: Option[Long] = (json \ "user" \ "name").asOpt[Long]
-maybeNameLong: Option[Long] = None
-'''
-
-<br/>
 ### Safest conversion with `validate[T]`

 `asOpt[T]` is better but you lose the kind of error that was detected.  
@@ -299,7 +175,7 @@
 - `JsResult[T]` accumulates all detected errors (doesn't stop at 1st error),
 - `JsResult[T]` is a monadic structure providing `map`/`flatMap`/`fold` operations to manipulate, compose it.

-#### `JsResult[T]` in a very nutshell
+#### `JsResult[T]` in a nutshell

 `JsResult[T]` can have 2 values:

@@ -314,162 +190,62 @@
 > `(JsPath \ "user" \ "name")(json)`
 > _You create a `JsPath` to search `user` then `name` and apply it to a given `json`._

-
 - `JsError(errors: Seq[(JsPath, Seq[ValidationError])])` :  
     - `errors` is a Sequence of pairs `(JsPath, Seq[ValidationError])`
     - pair `(JsPath, Seq[ValidationError])` locates one or more detected errors at given `JsPath`

-A few samples of usage:
+So a successful conversion might look like this:

-'''scala
-scala> import play.api.libs.json._
-
-scala> val jsres: JsResult[String] = JsString("toto").validate[String]
-jsres: JsSuccess("toto")
+@[validate-success](code/ScalaJson.scala)

-scala> val jsres: JsResult[String] = JsNumber(123).validate[String]
-jsres: play.api.libs.json.JsResult[String] = JsError(List((,List(ValidationError(validate.error.expected.jsstring,WrappedArray())))))
+If however the path wasn't found, we get this:

-jsres.map{ s: String => …}
-jsres.flatMap{ s: String => JsSuccess(s) }
+@[validate-failure](code/ScalaJson.scala)

-jsres.fold( 
-  errors: Seq[(JsPath, Seq[ValidationError])] => // manage errors,
-  s: String => // manage value 
-)
+Since `map` and `flatMap` are provided, for comprehensions can easily be used to extract values out:

-jsres.map( s: String => // manage value )
-     .recoverTotal( jserror: JsError => // manage errors and return default value)
-'''
+@[validate-compose](code/ScalaJson.scala)

-#### case OK: path found & conversion possible
-
-'''scala
-scala> val safeName = (json \ "user" \ "name").validate[String]
-safeName: play.api.libs.json.JsResult[String] = JsSuccess(toto,) // path is not precised because it's root
-'''
-
-#### case KO: Path not found
-
-'''scala
-scala> val nameXXX = (json \ "user" \ "nameXXX").validate[String]
-nameXXX: play.api.libs.json.JsResult[String] = 
-  JsError(List((,List(ValidationError(validate.error.expected.jsstring,WrappedArray())))))
-'''
-
-> Please note the error that doesn't return `path.not.found` as you may expect. This is a difference from JSON combinators presented later in the doc. 
-> This is due to the fact that `(json \ "user" \ "nameXXX")` returns `JsNull` and the implicit `Reads[String]` here awaits a `JsString` which explains the detected error.
-
-#### case KO: Conversion not possible
-
-'''scala
-scala> val name = (json \ "user" \ "name").validate[Long]
-name: play.api.libs.json.JsResult[Long] = 
-  JsError(List((,List(ValidationError(validate.error.expected.jsnumber,WrappedArray())))))
-'''
-
-<br/>
 ### Converting Recursive path `\\`

 `\\` recursively searches in the sub-tree and returns a `Seq[JsValue]` of found JsValue which is then a collection with classical Scala functions.

-'''scala
-scala> val emails: Seq[String] = (json \ "user" \\ "email").map(_.as[String])
-emails: Seq[String] = List(toto@jmail.com, tata@coldmail.com)
-'''
-<br/>
+@[recursive-as](code/ScalaJson.scala)
+
 ## Converting a Scala value to JsValue

 Scala to JSON conversion is performed by function `Json.toJson[T](implicit writes: Writes[T])` based on implicit typeclass `Writes[T]` which is just able to convert a `T` to a `JsValue`. 

 ### Create very simple JsValue

-'''
-val jsonNumber = Json.toJson(4)
-jsonNumber: play.api.libs.json.JsValue = 4
-'''
+@[convert-simple-type](code/ScalaJson.scala)

 *This conversion is possible because Play JSON API provides an implicit `Writes[Int]`*

+### Create a JSON array from a `Seq[T]`

-### Create a JSON array from a Seq[T]
-
-'''
-import play.api.libs.json.Json
-
-val jsonArray = Json.toJson(Seq(1, 2, 3, 4))
-jsonArray: play.api.libs.json.JsValue = [1,2,3,4]
-'''
+@[convert-seq](code/ScalaJson.scala)

 *This conversion is possible because Play JSON API provides an implicit `Writes[Seq[Int]]`*

 Here we have no problem to convert a `Seq[Int]` into a Json array. However it is more complicated if the `Seq` contains heterogeneous values:

-'''
-import play.api.libs.json.Json
+'''scala
+import play.api.libs.json._

 val jsonArray = Json.toJson(Seq(1, "Bob", 3, 4))
-<console>:11: error: No Json deserializer found for type Seq[Any]. Try to implement an implicit Writes or Format for this type.
-       val jsonArray = Json.toJson(Seq(1, "Bob", 3, 4))
-'''
-
-You get an error because there is no way to convert a `Seq[Any]` to Json (`Any` could be anything including something not supported by Json right?)
-
-A simple solution is to handle it as a `Seq[JsValue]`:
-
 '''
-import play.api.libs.json.Json

-val jsonArray = Json.toJson(Seq(
-  toJson(1), toJson("Bob"), toJson(3), toJson(4)
-))
-'''
-*This conversion is possible because Play API JSON provides an implicit `Writes[Seq[JsValue]]`*
+When we try to compile this, we get the following error:

-### Create a JSON object from a Map[String, T]
+    No Json deserializer found for type Seq[Any]. Try to implement an implicit Writes or Format for this type.

-'''
-import play.api.libs.json.Json
+This is because there is no way to convert a `Seq[Any]` to Json (`Any` could be anything including something not supported by Json right?)

-val jsonObject = Json.toJson(
-  Map(
-    "users" -> Seq(
-      toJson(
-        Map(
-          "name" -> toJson("Bob"),
-          "age" -> toJson(31),
-          "email" -> toJson("bob@gmail.com")
-        )
-      ),
-      toJson(
-        Map(
-          "name" -> toJson("Kiki"),
-          "age" -> toJson(25),
-          "email" -> JsNull
-        )
-      )
-    )
-  )
-)
-'''
+A simple solution is to handle it as a `Seq[JsValue]`:

-That will generate this Json result:
+@[convert-hetro-seq](code/ScalaJson.scala)

-'''
-{
-  "users":[
-    {
-      "name": "Bob",
-      "age": 31.0,
-      "email": "bob@gmail.com"
-    },
-    {
-      "name": "Kiki",
-      "age":  25.0,
-      "email": null
-    }
-  ]
-}
-'''
+*This conversion is possible because Play API JSON provides an implicit `Writes[Seq[JsValue]]`*

 > **Next:** [[JSON Reads/Writes/Formats Combinators | ScalaJsonCombinators]]
\ No newline at end of file

```

**Comment by garbagetown** [comment]
pull reqed at #692

**Comment by wm3** [comment]
merged #692
